tests/failpoints/cases/test_cmd_epoch_checker.rs

```diff
@@ -0,0 +1,368 @@
+    let (proposed_tx, proposed_rx) = mpsc::channel();
+    let (committed_tx, committed_rx) = mpsc::channel();
+    let (cb, applied_rx) = make_cb_ext(
+        cmd,
+        Some(Box::new(move || proposed_tx.send(()).unwrap())),
+        Some(Box::new(move || committed_tx.send(()).unwrap())),
+    );
+    (
+        cb,
+        CbReceivers {
+            proposed: proposed_rx,
+            committed: committed_rx,
+            applied: applied_rx,
+        },
+    )
+}
+
+fn make_write_req(cluster: &mut Cluster<NodeCluster>, k: &[u8]) -> RaftCmdRequest {
+    let r = cluster.get_region(k);
+    let mut req = new_request(
+        r.get_id(),
+        r.get_region_epoch().clone(),
+        vec![new_put_cmd(k, b"v")],
+        false,
+    );
+    let leader = cluster.leader_of_region(r.get_id()).unwrap();
+    req.mut_header().set_peer(leader);
+    req
+}
+
+#[test]
+fn test_reject_proposal_during_region_split() {
+    let mut cluster = new_node_cluster(0, 3);
+    let pd_client = cluster.pd_client.clone();
+    pd_client.disable_default_operator();
+    cluster.run();
+    cluster.must_transfer_leader(1, new_peer(3, 3));
+    cluster.must_put(b"k", b"v");
+
+    // Pause on applying so that region split is not finished.
+    let fp = "apply_before_split_1_3";
+    // let fp = "apply_before_split";
+    fail::cfg(fp, "pause").unwrap();
+
+    // Try to split region.
+    let (split_tx, split_rx) = mpsc::channel();
+    let cb = Callback::Read(Box::new(move |resp: ReadResponse<RocksSnapshot>| {
+        split_tx.send(resp.response).unwrap()
+    }));
+    let r = cluster.get_region(b"");
+    cluster.split_region(&r, b"k", cb);
+    split_rx
+        .recv_timeout(Duration::from_millis(100))
+        .unwrap_err();
+
+    // Try to put a key.
+    let write_req = make_write_req(&mut cluster, b"k1");
+    let (cb, cb_receivers) = make_cb(&write_req);
+    cluster
+        .sim
+        .rl()
+        .async_command_on_node(3, write_req, cb)
+        .unwrap();
+    // The write request should be blocked until split is finished.
+    cb_receivers.assert_not_ready();
+
+    fail::remove(fp);
+    // Split is finished.
+    assert!(!split_rx
+        .recv_timeout(Duration::from_secs(1))
+        .unwrap()
+        .get_header()
+        .has_error());
+
+    // The write request fails due to epoch not match.
+    cb_receivers.assert_err();
+
+    // New write request can succeed.
+    let write_req = make_write_req(&mut cluster, b"k1");
+    let (cb, cb_receivers) = make_cb(&write_req);
+    cluster
+        .sim
+        .rl()
+        .async_command_on_node(3, write_req, cb)
+        .unwrap();
+    cb_receivers.assert_ok();
+}
+
+#[test]
+fn test_reject_proposal_during_region_merge() {
+    let mut cluster = new_node_cluster(0, 3);
+    configure_for_merge(&mut cluster);
+    let pd_client = cluster.pd_client.clone();
+    pd_client.disable_default_operator();
+    cluster.run();
+    cluster.must_transfer_leader(1, new_peer(1, 1));
```

Comment:
Seems it's not necessary.

Reply:
It's necessary. I have to know which node is the leader. `run` can't guarantee node-1 is the leader.